File: src/main/java/ru/timmson/kanban/game/model/CardFactory.java
package ru.timmson.kanban.game.model;

import java.util.Arrays;
import java.util.List;
import java.util.function.IntFunction;

import static ru.timmson.kanban.game.model.Card.*;
import static ru.timmson.kanban.game.model.Stage.*;

public class CardFactory {

    //////////////////////Cards///////////////////////////////////

    public static Card F(Integer id, IntFunction<Integer> value,  Work... works) {
        return createCard(id, CARD_TYPE_F, value, Arrays.asList(works));
    }

    public static Card I(Integer id, IntFunction<Integer> value, Work... works) {
        return createCard(id, CARD_TYPE_I, value, Arrays.asList(works));
    }

    public static Card E(Integer id, IntFunction<Integer> value, Work... works) {
        return createCard(id, CARD_TYPE_E, value, Arrays.asList(works));
    }

    public static Card S(Integer id, IntFunction<Integer> value, Work... works) {
        return createCard(id, CARD_TYPE_S, value, Arrays.asList(works));
    }

    private static Card createCard(Integer id, String cardType, IntFunction<Integer> value, List<Work> works) {
        return new Card(id, cardType, value, works);
    }


    //////////////////////Works///////////////////////////////////

    public static Work A(Integer points) {
        return createWork(STAGE_ANALYSIS, points);
    }

    public static Work D(Integer points) {
        return createWork(STAGE_DEVELOPMENT, points);
    }

    public static Work T(Integer points) {
        return createWork(STAGE_TESTING, points);
    }

    private static Work createWork(Stage stage, Integer points) {
        return new Work(stage,points);
    }


    //////////////////////Dice///////////////////////////////////

    public static Dice DICE(DiceSide... diceSides) {
        return new Dice(Arrays.asList(diceSides));
    }

    public static DiceSide SIDE(Work... works) {
        return new DiceSide(Arrays.asList(works));
    }
}

File: src/main/kotlin/ru/timmson/kanban/game/model/CardFactory.kt
package ru.timmson.kanban.game.model

import ru.timmson.kanban.game.model.Card.Companion.CARD_TYPE_E
import ru.timmson.kanban.game.model.Card.Companion.CARD_TYPE_F
import ru.timmson.kanban.game.model.Card.Companion.CARD_TYPE_I
import ru.timmson.kanban.game.model.Card.Companion.CARD_TYPE_S
import ru.timmson.kanban.game.model.Stage.Companion.STAGE_ANALYSIS
import ru.timmson.kanban.game.model.Stage.Companion.STAGE_DEVELOPMENT
import ru.timmson.kanban.game.model.Stage.Companion.STAGE_TESTING
import java.util.*

class CardFactory {
    companion object {
        @JvmStatic
        fun f(id: Int?, value: (Int) -> Int, vararg works: Work): Card {
            return createCard(id, CARD_TYPE_F, value, *works)
        }

        @JvmStatic
        fun i(id: Int?, value: (Int) -> Int, vararg works: Work): Card {
            return createCard(id, CARD_TYPE_I, value, *works)
        }

        @JvmStatic
        fun e(id: Int?, value: (Int) -> Int, vararg works: Work): Card {
            return createCard(id, CARD_TYPE_E, value, *works)
        }

        @JvmStatic
        fun s(id: Int?, value: (Int) -> Int, vararg works: Work): Card {
            return createCard(id, CARD_TYPE_S, value, *works)
        }

        private fun createCard(id: Int?, cardType: String, value: (Int) -> Int, vararg works: Work): Card {
            return Card(id!!, cardType, value, listOf(*works))
        }

        @JvmStatic
        fun a(points: Int?): Work {
            return createWork(STAGE_ANALYSIS, points)
        }

        @JvmStatic
        fun d(points: Int?): Work {
            return createWork(STAGE_DEVELOPMENT, points)
        }

        @JvmStatic
        fun t(points: Int?): Work {
            return createWork(STAGE_TESTING, points)
        }

        private fun createWork(stage: Stage, points: Int?): Work {
            return Work(stage, points)
        }

        @JvmStatic
        fun dice(vararg diceSides: DiceSide): Dice {
            return Dice(Arrays.asList(*diceSides))
        }

        @JvmStatic
        fun side(vararg works: Work): DiceSide {
            return DiceSide(Arrays.asList(*works))
        }
    }
}